kafka-connect/src/main/java/io/tabular/iceberg/connect/channel/ExceptionHandlingTableWriter.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package io.tabular.iceberg.connect.channel;
+
+import io.tabular.iceberg.connect.WriteExceptionHandler;
+import io.tabular.iceberg.connect.data.Utilities;
+import io.tabular.iceberg.connect.data.WriterResult;
+import java.util.List;
+import org.apache.kafka.connect.sink.SinkRecord;
+
+class ExceptionHandlingTableWriter implements TableWriter, AutoCloseable {
+
+  private final TableWriter underlying;
+  private final WriteExceptionHandler exceptionHandler;
+
+  ExceptionHandlingTableWriter(TableWriter underlying, WriteExceptionHandler exceptionHandler) {
+    this.underlying = underlying;
+    this.exceptionHandler = exceptionHandler;
+  }
+
+  @Override
+  public void write(SinkRecord sinkRecord, String tableName, boolean ignoreMissingTable) {
+    try {
+      underlying.write(sinkRecord, tableName, ignoreMissingTable);
```

Comment:
This will also cause infinite loops if you fail to create the dead letter table (possibly because the dead letter route has been set to an invalid value).  Also an issue that this breaks due to whether or not autocreate tables is configured on. 

Again, this is why CatalogAPI exists in the other API: 

- it knows when you are creating / doing something with the dead letter table, so throw those errors instead of attempting to throw them into a dead letter table and do not catch the exception
- when you are doing something with a regular record/not dead letter table, then apply the error handler.

I can still do the above w/ the error handler, but you can't have this broad `write` here --too many things downstream, some of which involve hard coded configs, some of which will involve the dead letter table --e.g. we support partitioning the dead letter table via the normal way to provide partition specs for any table, if that fails, you need to hard fail the connector.  

The above would infinite loop in that case.

Reply:
> This will also cause infinite loops

I deliberately wrote it recursively. If users (which could be us, the maintainers!) feel an exception is transient, they're welcome to retry via the API, I don't feel we should restrict that option. 

> possibly because the dead letter route has been set to an invalid value

There is no dead-letter-table-route with this approach as far as the connector code is concerned. 
The WriteExceptionHandler may certainly try to write to a dead-letter-table but that is a WriteExceptionHandler implementation detail. 

> I can still do the above w/ the error handler, but you can't have this broad write here --too many things downstream, some of which involve hard coded configs, some of which will involve the dead letter table --e.g. we support partitioning the dead letter table via the normal way to provide partition specs for any table, if that fails, you need to hard fail the connector.

With this approach, a dead-letter-table is like any other ordinary table so you can still "partitioning the dead letter table via the normal way to provide partition specs for any table." 

> if that fails, you need to hard fail the connector.

No problems, you can still achieve that here with the `WriteExceptionHandler` approach. The `WriteExceptionHandler` implementation just needs to ensure that if it sees a `PartitionSpecEvolutionError` for what it considers to be a dead-letter-table, it should just hard-fail :) 

> Again, this is why CatalogAPI exists in the other API:

Like I said, I'm not opposed to having a concept like the Catalog API in the connector code if it makes it easier to react to clear and actionable exceptions.